### DISCUSSION TITLE
EF test can be improved in order not to check same URLs multiple times

### DISCUSSION THREAD

**Discussion by iuriemaxim** [INSPIRE-MIF/helpdesk-validator#783] [Ideas]
The suggestion is to improve the algorithm used by the validator when testing data based on Environmental Monitoring Facilities (EF) Data Theme.

By validating the https://inspire-staging.meteoromania.ro/WIGOS/WFS?service=wfs&version=2.0.0&request=GetFeature&typeNames=ef:EnvironmentalMonitoringFacility
against the EF data specification, it was found that URLs are verified multiple times.

Below some examples of URLs that are verified multiple times (the number in front indicates how many times those URLs were checked).

46 | https://inspire.meteoromania.ro/observedProperty/BUFR0-13-023
23 | https://inspire.meteoromania.ro/observedProperty/BUFR0-11-001SF32PT10M
23 | https://inspire.meteoromania.ro/observedProperty/BUFR0-11-002SF32PT10M
23 | https://inspire.meteoromania.ro/observedProperty/BUFR0-12-023
23 | https://inspire.meteoromania.ro/observedProperty/BUFR0-12-023SF2PT24H
23 | https://inspire.meteoromania.ro/observedProperty/BUFR0-12-023SF3PT24H
23 | https://inspire.meteoromania.ro/observedProperty/BUFR0-12-023SF4PT24H
23 | https://inspire.meteoromania.ro/observedProperty/BUFR0-13-003
23 | https://inspire.meteoromania.ro/process/DailyAverageAirTemperatureFrom4Values
23 | https://inspire.meteoromania.ro/process/DailyMaximumAirTemperatureFrom4Values
23 | https://inspire.meteoromania.ro/process/DailyMinimumAirTemperatureFrom4Values
23 | https://inspire.meteoromania.ro/process/RelativeHumidity
23 | https://inspire.meteoromania.ro/process/Temperature
23 | https://inspire.meteoromania.ro/process/TotalPrecipitations24HoursFrom0530UTC
23 | https://inspire.meteoromania.ro/process/TotalPrecipitations24HoursFrom1730UTC
23 | https://inspire.meteoromania.ro/process/WindDirection
23 | https://inspire.meteoromania.ro/process/WindSpeed
9 | https://inspire.meteoromania.ro/ids/EnvironmentalMonitoringFacility.15015
9 | https://inspire.meteoromania.ro/ids/EnvironmentalMonitoringFacility.15020
9 | https://inspire.meteoromania.ro/ids/EnvironmentalMonitoringFacility.15090
9 | https://inspire.meteoromania.ro/ids/EnvironmentalMonitoringFacility.15108

Taking in consideration the specifics of EF data theme it is critical not to verify the same URL more than one time.
For a full implementation with many EnvironmentalMonitoringFacilities that will monitor multiple parameters (observed properties) based on multiple processes, the numbers will raise to hundreds instead of tens as it is now, so it will become critical to check a URL only once and not again, again and again.

By looking in the test log, I see that the URL for the EPSG code, namely

http://www.opengis.net/def/crs/EPSG/0/4258 

is checked only once, even if it appears many, many times.

07.10.2020 17:18:34 - "Checking URL: 'http://www.opengis.net/def/crs/EPSG/0/4258'"

So a similar approach as for EPSG URLs should be used for all URLs, in order to check them only once.

So, the ETF is fine, but should be optimized in order not to check the same URL several times.

Some more details can be found in https://github.com/inspire-eu-validation/community/issues/399

**Comment by carlospzurita** [comment]
Dear @iuriemaxim 

Thank you for testing this the Environmental Facilities and sharing your feedback, is very much appreciated.

You're right that making repeated requests to these services is not the most efficient. However, the ETS is designed in this way to preserve the tests agnostics from any service or data that may come. This is way, whenever an external item is referenced, an HTTP call is performed, and avoid defining an endless list of variables to hold different responses for schemas and service requests.

Nonetheless, and to boos the performance in this situations, the validator has integrated on the deployment level a cache system, to hold for a given amount of time responses of each request. Even if the ETF is performing the HTTP(S) requests, these are first intercepted by the cache before sending them to the service, speeding up the process.

**Comment by iuriemaxim** [comment]
@carlospzurita Thank you very much for the valuable information. I will have a look in the server logs as well.

**Comment by iuriemaxim** [comment]
Dear @carlospzurita 

This is the log of the requests received by the WFS:
[20201003_Log_Geo1.xlsx](https://github.com/inspire-eu-validation/community/files/5372023/20201003_Log_Geo1.xlsx)

Luckily the system is responding quite fast, as we also implemented a caching mechanism.

The log results looks like this, being quite similar with the log of the INSPIRE validator:

![image](https://user-images.githubusercontent.com/20813598/95874493-9b248d80-0d79-11eb-9805-c8140c1910b3.png)

If a new test is necessary to verify the functionality of the ETF caching system, please let me know.

**Comment by iuriemaxim** [comment]
@fabiovinci Hope that this enhancement will be developed in 2022.

**Comment by fabiovinci** [comment]
Dear @iuriemaxim,

in the frame of the new contract "_Operational support to the maintenance and technical evolution of INSPIRE components and artefacts_", we are reviewing all the open issues and prioritizing them in relation to their severity.

Since it is not strictly an issue, I will convert it into a discussion.
